File: app/src/main/java/com/originb/inkwisenote2/modules/backgroundjobs/BackgroundOps.java
package com.originb.inkwisenote2.modules.backgroundjobs;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;

import java.util.Optional;
import java.util.concurrent.*;
import java.util.function.Consumer;

public class BackgroundOps {

    private static BackgroundOps instance;
    private final Handler mainThreadHandler;

    private ExecutorService executor = Executors.newFixedThreadPool(10);

    public static BackgroundOps getInstance() {
        if (instance == null) {
            instance = new BackgroundOps();
        }
        return instance;
    }

    private BackgroundOps() {
        mainThreadHandler = new Handler(Looper.getMainLooper());
    }

    public static void execute(Runnable runnable) {
        getInstance().executor.execute(runnable);
    }

    public static void execute(Runnable runnable, Runnable continueOnMainThread) {
        try {
            getInstance().executor.execute(runnable);
            getInstance().mainThreadHandler.post(continueOnMainThread::run);

        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static <T> void execute(Callable<T> callable, Consumer<T> resultOnMainThread) {
        // 1. Capture the stack trace of the CALLER (the UI thread) right now
        Exception clientStack = new Exception("Task submitted from here");

        getInstance().executor.execute(() -> {
            try {
                T result = callable.call();
                getInstance().mainThreadHandler.post(() -> resultOnMainThread.accept(result));
            } catch (Exception e) {
                // 2. Stitch the current background exception with the caller's stack trace
                e.initCause(clientStack);

                // 3. Now when you log this, you see exactly where it was enqueued
                handleError(e);
            }
        });
    }

    public static <T> void executeOpt(Callable<Optional<T>> callable, Consumer<T> resultOnMainThread) {
        getInstance().executor.execute(() -> {
            try {
                Optional<T> result = callable.call();
                if (result.isPresent()) {
                    getInstance().mainThreadHandler.post(() -> resultOnMainThread.accept(result.get()));
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
    }

    private static void handleError(Exception e) {
        // Using a logger helps preserve the full cause chain
        e.printStackTrace();
        Log.e("BackgroundOps", "Background task failed", e);

        // Optional: Post back to UI to show an error dialog/toast
        getInstance().mainThreadHandler.post(() -> {
            // You could add an Error Consumer to your execute method if needed
        });
    }
}

File: app/src/main/java/com/originb/inkwisenote2/modules/backgroundjobs/BackgroundOps.kt
package com.originb.inkwisenote2.modules.backgroundjobs

import android.os.Handler
import android.os.Looper
import java.util.*
import java.util.concurrent.Callable
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.function.Consumer

class BackgroundOps private constructor() {
    private val mainThreadHandler = Handler(Looper.getMainLooper())

    private val executor: ExecutorService = Executors.newFixedThreadPool(10)

    companion object {
        var instance: BackgroundOps? = null
            get() {
                if (field == null) {
                    field = BackgroundOps()
                }
                return field
            }
            private set

        fun execute(runnable: Runnable?) {
            instance!!.executor.execute(runnable)
        }

        fun execute(runnable: Runnable?, continueOnMainThread: Runnable) {
            try {
                instance!!.executor.execute(runnable)
                instance!!.mainThreadHandler.post { continueOnMainThread.run() }
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
        }

        fun <T> execute(callable: Callable<T>?, resultOnMainThread: Consumer<T>) {
            try {
                val callFuture = instance!!.executor.submit(callable)
                val result = callFuture.get()

                instance!!.mainThreadHandler.post { resultOnMainThread.accept(result) }
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
        }

        fun <T> executeOpt(callable: Callable<Optional<T>?>?, resultOnMainThread: Consumer<T>) {
            try {
                val callFuture = instance!!.executor.submit(callable)
                val result = callFuture.get()

                if (result!!.isPresent) {
                    instance!!.mainThreadHandler.post {
                        resultOnMainThread.accept(
                            result.get()
                        )
                    }
                }
            } catch (e: Exception) {
                throw RuntimeException(e)
            }
        }
    }
}
